Add Permissions framework and definitions.

DarthLovegood opened the pull request:
This initial commit only enforces permission-checking in the Permissions and Help cogs, to avoid touching too many files at once.

DarthLovegood committed:
Add Permissions framework and definitions.

This initial commit only enforces permission-checking in the Permissions and Help cogs, to avoid touching too many files at once.

ThisIsCad commented on cogs/permissions.py:
```diff
@@ -0,0 +1,439 @@
+            async with connect(self.db) as connection:
+                is_enabled = permission_config.is_enabled
+                whitelisted_channel_ids = dumps(sorted(permission_config.whitelisted_channel_ids))
+                query = 'SELECT * FROM permissions WHERE server_id=? AND permission_id=?'
```
Hmmm this may be more concisely represented with an [upsert](https://www.sqlite.org/lang_UPSERT.html) operator

ThisIsCad commented on cogs/permissions.py:
```diff
@@ -0,0 +1,439 @@
+
+            async with connect(self.db) as connection:
+                from_statement = 'FROM permissions WHERE server_id=? AND permission_id=?'
+                async with connection.execute(f'SELECT * {from_statement}', (server_id, permission.id)) as cursor:
```
I see no reason for this check, sqlite will just return the number of affected rows (e.g. 0 or 1) and will not throw any warnings/errors if the row doesnt exist

DarthLovegood committed:
Clean up SQL, handle unavailable channels, fix some logging.

DarthLovegood commented:
Cleaned up SQL according to review comments, added some handling for unavailable channels (i.e. channels where the bot doesn't have permission to send messages), and fixed some logging.
